File: 4_data_processing/filter.jl
#using SpineInterface
using JSON

#path = dirname(@__DIR__)
genericenergyformat = ARGS[1]#joinpath(path, "0_component_tool_template", "generic_data_format.json")#only generic format
usersettings = ARGS[2]#joinpath(path, "4_data_processing", "usersettings.json")#only entities
dataset = ARGS[3]#joinpath(path, "3_select_unprocessed_data", "fulldetaildata.json")#only parameters
casedata = ARGS[4]#joinpath(path, "5_built_data_and_scenarios", "casedata.json")#

# improvement: use the database operations instead of dictionaries
# improvement: merging user and dataset for both entities and parameters

# load generic energy format
println("Load generic energy format")
iodb = open(genericenergyformat, "r") do f
	dicttxt = read(f,String) # file information to string
	return JSON.parse(dicttxt) # parse and transform data
end

# make convenience parameter dictionary
parameterdefinition = Dict()
for parameter in iodb["parameter_definitions"]
	push!(get!(parameterdefinition, parameter[1], []), parameter[2])
end

# load user specifications
println("Load user database")
userdata =  open(usersettings, "r") do f
	dicttxt = read(f,String) # file information to string
	return JSON.parse(dicttxt) # parse and transform data
end
userentities = userdata["entities"]

# make convenience parameter dictionary

# load full dataset
println("Load full database")
#using_spinedb(dataset)
fulldata = open(dataset, "r") do f
	dicttxt = read(f,String)
	return JSON.parse(dicttxt)
end

# create case data
println("Stack the template with a part of the full database and the user database")
alternatives = []
for userentity in userentities
	# entities
	push!(iodb["entities"], userentity)
	# parameters
	userentityname = userentity[2]
	if size(userentity[3],1)>1
		userentityrelation = userentity[3]
	else
		userentityrelation = userentityname
	end
	try#should fail because SpineInterface is disabled
		entity = getfield(Main,Symbol(userentityname))()
		for parametername in parameterdefinition[entity.class_name]
			kwarg = Dict(Symbol(entity.class_name)=>entity)
			parametervalue = getfield(Main,Symbol(parametername))(;kwarg...)
			push!(iodb["parameter_values"],[entity.class_name,userentityrelation,parameter,parametervalue,"Base"])
		end
	catch
		#do nothing
	end
	#use dicts as long as SpineInterface is disabled
	#entity = fulldata[userentityname]
	for parametername in parameterdefinition[userentity[1]]
		entityparametervalue = nothing
		for parametervalue in fulldata["parameter_values"]
			if parametername == parametervalue[3] && userentityrelation == parametervalue[2]
				entityparametervalue = parametervalue
			end
		end
		for parametervalue in userdata["parameter_values"]
			if parametername == parametervalue[3] && userentityrelation == parametervalue[2]
				entityparametervalue = parametervalue
			end
		end
		if isnothing(entityparametervalue) == false
			push!(iodb["parameter_values"],entityparametervalue)
		end
	end
end
for alternative in alternatives
	push!(iodb["alternatives"], [alternative, ""])
end

#bring the data to the select data set
println("Create Case data")
#import_data(casedata, iodb, "Generate Case data")
open(casedata, "w") do f
    JSON.print(f, iodb, 4)
end

File: ines_tools/filter.py
import sys
import json
import spinedb_api as api
from spinedb_api import purge

ARGS = sys.argv[1:]
ines = ARGS[0]
usersettings = ARGS[1]
dataset = ARGS[2]
casedata = ARGS[3]

# improvement: use the database operations instead of dictionaries
# improvement: merging user and dataset for both entities and parameters

# load generic energy format
print("Load ines spec")
with open(ines) as f:
	iodb = json.load(f)
	#the new version of the ines spec is missing fields that the script relies on:
	iodb["entities"]=[]
	iodb["parameter_values"]=[]
	iodb["alternatives"]=[]

# make convenience parameter dictionary
parameterdefinition = {}
for parameter in iodb["parameter_definitions"]:
	if parameter[0] not in parameterdefinition:
		parameterdefinition[parameter[0]] = []
	parameterdefinition[parameter[0]].append(parameter[1])

# load user specifications
print("Load user database")
with open(usersettings) as f:
	userdata = json.load(f)
userentities = userdata["entities"]

# make convenience parameter dictionary

# load full dataset
print("Load full database")
if dataset.split(".")[-1] == 'json':
	with open(dataset) as f:
		fulldata = json.load(f)
else:
	with api.DatabaseMapping(dataset) as source_db:
		fulldata = api.export_data(source_db,parse_value=api.parameter_value.load_db_value)

# create case data
print("Stack the template with a part of the full database and the user database")
alternatives = []
for userentity in userentities:
	# entities
	iodb["entities"].append(userentity)
	# parameters
	userentityclass = userentity[0]
	userentityname = userentity[1]
	#entity = fulldata[userentityname]
	for parametername in parameterdefinition[userentityclass]:
		entityparametervalue = None
		for parametervalue in fulldata["parameter_values"]:
			if parametername == parametervalue[2] and userentityname == parametervalue[1]:
				entityparametervalue = parametervalue
		for parametervalue in userdata["parameter_values"]:
			if parametername == parametervalue[2] and userentityname == parametervalue[1]:
				entityparametervalue = parametervalue
		if entityparametervalue != None:
			iodb["parameter_values"].append(entityparametervalue)
			if entityparametervalue[4] not in alternatives:
				alternatives.append(entityparametervalue[4])

for alternative in alternatives:
	iodb["alternatives"].append([alternative, ""])

#bring the data to the select data set
print("Create Case data")
#import_data(casedata, iodb, "Generate Case data")
if casedata.split(".")[-1] == 'json':
    with open(casedata, 'w') as f:
        json.dump(iodb, f, indent=4)
else:
	with api.DatabaseMapping(casedata) as target_db:
		purge.purge(target_db, purge_settings=None)
		message = api.import_data(target_db,**iodb)
		target_db.refresh_session()
		target_db.commit_session("Import data")

		# debugging
		#print(message)